src/core/ensemble_scheduler.h

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <memory>
+#include "src/core/model_config.pb.h"
+#include "src/core/model_config_utils.h"
+#include "src/core/provider.h"
+#include "src/core/scheduler.h"
+#include "src/core/status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+class InferenceServer;
+
+// Scheduler that implements ensemble scheduling.
+class EnsembleScheduler : public Scheduler {
+ public:
+  // Create a scheduler to process ensemble requests and
+  // to dispatch requests to models in ensemble internally.
+  static Status Create(
+      const ModelConfig& config, std::unique_ptr<Scheduler>* scheduler);
+
+  // \see Scheduler::Enqueue()
+  void Enqueue(
+      const std::shared_ptr<ModelInferStats>& stats,
+      const std::shared_ptr<InferRequestProvider>& request_provider,
+      const std::shared_ptr<InferResponseProvider>& response_provider,
+      std::function<void(Status)> OnComplete) override;
+
+  // Set the inference server that the scheduler is communicating with
+  Status SetInferenceServer(void* inference_server)
```

Comment:
Override

Reply:
Override is still missing on SetInferenceServer

Reply:
This is a new function for EnsembleScheduler, the Scheduler interface doesn't have this function. Should I also add this to Scheduler and default to do nothing?

Reply:
My mistake, I thought it was the override from Backend.